### ISSUE TITLE
[BUG] 16 modules fail to build on the latest master

### ISSUE THREAD

**Issue by hammady** [bitfocus/companion#1937]
### Is this a bug in companion itself or a module?

- [X] I believe this to be a bug in companion

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

There are 16 modules that fail the yarn command due to compilation errors in the JS code. They may require upgrading. I am not sure if this is related to a new `instance_skel` or ES6/TS. The list is below. Sorry for not posting the issue on those repos because this is a lot and I tend to belive it has something to do with the core module code.

```
"companion-module-barco-clickshare": "github:bitfocus/companion-module-barco-clickshare#v1.0.0",
"companion-module-behringer-x32": "github:bitfocus/companion-module-behringer-x32#v2.9.1",
"companion-module-bitfocus-snapshot": "github:bitfocus/companion-module-bitfocus-snapshot#v0.0.5",
"companion-module-bmd-atem": "github:bitfocus/companion-module-bmd-atem#v2.13.3",
"companion-module-cisco-webex-websocket": "github:bitfocus/companion-module-cisco-webex-websocket#v1.1.0",
"companion-module-generic-emberplus": "github:bitfocus/companion-module-generic-emberplus#v1.1.4",
"companion-module-homeassistant-server": "github:bitfocus/companion-module-homeassistant-server#v0.7.1",
"companion-module-iiyama-prolite":"github:bitfocus/companion-module-iiyama-prolite#v1.0.0",
"companion-module-magewell-ultrastream": "github:bitfocus/companion-module-magewell-ultrastream#v0.0.4",
"companion-module-seervision-suite": "github:bitfocus/companion-module-seervision-suite#v1.2.1",
"companion-module-sonos-speakers": "github:bitfocus/companion-module-sonos-speakers#v0.2.1",
"companion-module-soundcraft-ui": "github:bitfocus/companion-module-soundcraft-ui#v2.0.5",
"companion-module-studiocoast-vmix": "github:bitfocus/companion-module-studiocoast-vmix#v2.0.7",
"companion-module-twitch-api": "github:bitfocus/companion-module-twitch-api#v2.0.4",
"companion-module-vystem-api": "github:bitfocus/companion-module-vystem-api#v1.0.2",
"companion-module-youtube-live": "github:bitfocus/companion-module-youtube-live#v1.1.5",
```

### Steps To Reproduce

1. Build from docker
2. Observe build errors
3. Remove the abovementioned modules
4. Build succeeds

### Expected Behavior

No build issues while including those 16 modules.

### Environment (please complete the following information)

```markdown
- OS: MacOS 12.0.1
- Browser: N/A
- Companion Version: 2.2 beta (master commit 762a6e94892e053dc56e9f2d339d1f746428a0a7).
```

**Comment by Julusian** [comment]
>  Sorry for not posting the issue on those repos because this is a lot and I tend to belive it has something to do with the core module code.

No, I think that here is best because that looks to be every typescript module, and only the typescript modules. What is it saying about them? I was trying master last night and it was perfectly happy with them all..

**Comment by hammady** [comment]
Well, you are right, on master/beta there are no build issues for the typescript modules using the Dockerfile. I was optimizing the dockerfile by copying parts of the code necessary for compilation then copy the rest after that to make use of caching (very convenient in development, but does not matter in production). I was copy only the below, so I am sure I missed something that caused the errors. Here are the errors anyway for one of them:

```
#17 45.68 Directory: /app/node_modules/companion-module-bitfocus-snapshot                                                                                                                                                                      
#17 45.68 Output:                                                                                                                                                                                                                              
#17 45.68 yarn run v1.22.17                                                                                                                                                                                                                    
#17 45.68 $ npx rimraf dist && npx typescript@~4.2 -p tsconfig.build.json                                                                                                                                                                      
#17 45.68 npx: installed 12 in 2.772s                                                                                                                                                                                                          
#17 45.68 npx: installed 1 in 3.585s                                                                                                                                                                                                           
#17 45.68 src/actions.ts:6:8 - error TS2307: Cannot find module '../../../instance_skel_types' or its corresponding type declarations.                                                                                                         
#17 45.68                                                                                                                                                                                                                                      
#17 45.68 6 } from '../../../instance_skel_types'                                                                                                                                                                                              
#17 45.68          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                                              
#17 45.68                                                                                                                                                                                                                                      
#17 45.68 src/config.ts:1:31 - error TS2307: Cannot find module '../../../instance_skel' or its corresponding type declarations.                                                                                                               
#17 45.68                                                                                                                                                                                                                                      
#17 45.68 1 import InstanceSkel = require('../../../instance_skel')                                                                                                                                                                            
#17 45.68                                 ~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                             
#17 45.68                                                                                                                                                                                                                                      
#17 45.68 src/config.ts:2:42 - error TS2307: Cannot find module '../../../instance_skel_types' or its corresponding type declarations.                                                                                                         
#17 45.68                                                                                                                                                                                                                                      
#17 45.68 2 import { SomeCompanionConfigField } from '../../../instance_skel_types'                                                                                                                                                            
#17 45.68                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                            
#17 45.68                                                                                                                                                                                                                                      
#17 45.68 src/index.ts:1:31 - error TS2307: Cannot find module '../../../instance_skel' or its corresponding type declarations.                                                                                                                
#17 45.68                                                                                                                                                                                                                                      
#17 45.68 1 import InstanceSkel = require('../../../instance_skel')                                                                                                                                                                            
#17 45.68                                 ~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                             
#17 45.68                                                                                                                                                                                                                                      
#17 45.68 src/index.ts:8:8 - error TS2307: Cannot find module '../../../instance_skel_types' or its corresponding type declarations.                                                                                                           
#17 45.68                                                                                                                                                                                                                                      
#17 45.68 8 } from '../../../instance_skel_types'                                                                                                                                                                                              
#17 45.68          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                                              
#17 45.68                                                                                                                                                                                                                                      
#17 45.68 src/index.ts:23:12 - error TS2339: Property 'config' does not exist on type 'SnapshotInstance'.                                                                                                                                      
#17 45.68                                                                                                                                                                                                                                      
#17 45.68 23   if (this.config.host && this.config.port && this.config.secret) {                                                                                                                                                               
#17 45.68               ~~~~~~                                                                                                                                                                                                                 
```

And here are the things I copy before the yarn:

```dockerfile
COPY tools /app/tools
COPY package.json /app/package.json
COPY yarn.lock /app/yarn.lock
COPY webui /app/webui
```

If developing the core companion inside docker is not something you are guys interested in, I can stop investing more time in it, although any pointers on what I could have missed that caused the above is very much appreciated.

**Comment by hammady** [comment]
OMG, I just had to look at the errors more closely to know what is missing!
By adding the below, it worked. Again, that's all an investment in docker development, what about a PR for a dev dockerfile to collect feedback?
```dockerfile
COPY *.ts /app/
```

**Comment by Julusian** [comment]
Ah, glad you figured it out :)

I think it depends on the purpose of the dev dockerfile. For module development or for core development?

If for modules, then you can use the existing dockerfile (or prebuilt images) and mount in a folder to module-local-dev and it should work. Perhaps overriding the entrypoint to get companion to auto-restart on file change.

If its for core development, I dont know what would be useful there. I need to understand the workflow to be able to form an opinion on how we can support it. I am hesitant to add another dockerfile, because its another thing to maintain and while my dev machine is running linux I do not use companion in docker at all right now, so it will inevitably end up breaking at points with no-one noticing (see the lack of docker image building for 29 days, and missing dependency)

I am open to changes to support either flow, but I need it to feel like it will be useful and not just be more to maintain

**Comment by hammady** [comment]
Sorry this slipped away from me. You are right, for module dev one can mount the module-local-dev and simply restart the container. I already do that in my setup. For core development, and for someone who wants to exclusively develop using docker to keep their host machines clean, the current Dockerfile will build all layers from scratch (taking minutes) because of the line: `COPY . /app/` as any file would invalidate the dot. Typically in development Dockerfiles, best practices dictate that system dependencies are installed first, then libraries, then the source code. This allows a developer to reuse the cached system dependencies and libraries and only rebuild from layers after that.
I did the necessary modifications to the Dockerfile, but had to copy files in stages to maximize the use of caching. I agree that it may add an overhead of an additional Dockerfile, but for docker exclusive development, it is inevitable, so it depends on the case that you want to support that or not.

```dockerfile
# dev.dockerfile
FROM node:14 as companion-builder

# User variable, define where to store the application config
ENV COMPANION_CONFIG_BASEDIR=/config

# Installation Prep
# libvips for sharp 0.29.0+ required for Apple M1
RUN apt-get update && apt-get install -y \
    libusb-1.0-0-dev \
    libudev-dev \
    unzip \
    cmake \
    && rm -rf /var/lib/apt/lists/*

WORKDIR /app
COPY tools /app/tools
COPY package.json /app/package.json
COPY yarn.lock /app/yarn.lock
COPY *.ts /app/
COPY webui /app/webui

# Install dependencies
RUN ./tools/yarn.sh

# strip back unnecessary dependencies
RUN yarn --prod

# Generate version number file
COPY .git /app/.git
RUN yarn build:writefile

# make the production image
FROM node:14-slim

WORKDIR /app
COPY / /app/
COPY --from=companion-builder /app/node_modules	/app/node_modules
COPY --from=companion-builder /app/webui	/app/webui
COPY --from=companion-builder /app/BUILD	/app/BUILD

# Delete the webui source
RUN mv webui/build webui-build \
    && rm -R webui \
    && mkdir webui \
    && mv webui-build webui/build

# cleanup up some stuff that shouldnt be preserved
RUN rm -R .git

# module-local-dev dependencies
# Mount when running if no dependencies are needed
# if there are dependencies, need to copy and build, later find a way to mount without touching dependencies
# VOLUME [ "/app/module-local-dev" ]

ENV COMPANION_CONFIG_BASEDIR /companion
RUN mkdir $COMPANION_CONFIG_BASEDIR && chown node:node $COMPANION_CONFIG_BASEDIR
USER node
EXPOSE 8000
ENV DEVELOPER 1
# Bind to 0.0.0.0, as access should be scoped down by how the port is exposed from docker
ENTRYPOINT ["./headless_ip.js", "0.0.0.0"]
```

For more information about this development approach, please have a look on [this](https://www.docker.com/blog/containerized-python-development-part-1) article, section: "Instruction order matters for leveraging build cache".

**Comment by hammady** [comment]
I'm not against using nodemon. In the above Dockerfile I only applied the concept of reordering layers in case a build is required but the development ready Dockerfile still needs work as you said. 
I didn't remove production related layers because again I only approached one concept and I wanted it to be compatible still with production. 

To answer your question on why need docker for development that's a long debate indeed 😅
In a nutshell, seasonal developers that contribute to many open source projects should not be taxed by doing specific system preparation for every single project. Nvm may have solved it for node version and dependencies, but not for system requirements (e.g apt install libusb). Nvm itself is a system dependency. What about other languages? If you think of it at a broader scale, docker guarantees a unified and consistent way of approaching almost any project without even knowing the language the code is written in! Isn't this just cool?
